adapters/teads/teads.go

```diff
@@ -0,0 +1,196 @@
+		teadsImpExt := &teadsImpExtension{
+			KV: teadsKV{
+				PlacementId: defaultImpExt.Bidder.PlacementId,
+			},
+		}
+		if extJson, err := json.Marshal(teadsImpExt); err != nil {
+			return &errortypes.BadInput{
+				Message: "Error stringify Imp.Ext object",
+			}
+		} else {
+			imp.Ext = extJson
+		}
+	}
+	return nil
+}
+
+// Builds enpoint url based on adapter-specific pub settings from imp.ext
+func (a *adapter) buildEndpointURL() (string, error) {
+	endpointParams := macros.EndpointTemplateParams{}
+	host, err := macros.ResolveMacros(a.endpointTemplate, endpointParams)
+
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Unable to parse endpoint url template: " + err.Error(),
+		}
+	}
+
+	thisURI, err := url.Parse(host)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Malformed URL: " + err.Error(),
+		}
+	}
+
+	return thisURI.String(), nil
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, _ *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
+	}
+
+	err := adapters.CheckResponseStatusCodeForErrors(response)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid))
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+
+			bidExtTeads, err := getTeadsRendererFromBidExt(bid.Ext)
+			if err != nil {
+				return nil, err
+			}
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid: &bid,
+				BidMeta: &openrtb_ext.ExtBidPrebidMeta{
+					RendererName:    bidExtTeads.Prebid.Meta.RendererName,
+					RendererVersion: bidExtTeads.Prebid.Meta.RendererVersion,
+				},
+				BidType: getMediaTypeForImp(bid.ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	if bidResp.Cur != "" {
+		bidderResponse.Currency = bidResp.Cur
+	}
+	return bidderResponse, nil
+}
+
+func getTeadsRendererFromBidExt(ext json.RawMessage) (*teadsBidExt, []error) {
+	var bidExtTeads teadsBidExt
+	if err := json.Unmarshal(ext, &bidExtTeads); err != nil {
+		return nil, []error{err}
+	}
+	if bidExtTeads.Prebid.Meta.RendererName == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "RendererName should not be empty if present",
+		}}
+	}
+	if bidExtTeads.Prebid.Meta.RendererVersion == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "RendererVersion should not be empty if present",
+		}}
+	}
+	return &bidExtTeads, nil
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo
+			}
+			return openrtb_ext.BidTypeBanner
+		}
+	}
+	return openrtb_ext.BidTypeBanner
```

Comment:
should error out if received `impID` that does not have any match in `imps`

Reply:
Updated, i could not use pointers because go does not allow me so i returned bidType `Banner` with error. If you have any tips on that

Reply:
could return `openrtb_ext.BidType("")` instead of `openrtb_ext.BidTypeBanner`

refer https://github.com/prebid/prebid-server/pull/3112#discussion_r1349904142